launcher/ui/dialogs/skins/SkinManageDialog.cpp

```diff
@@ -0,0 +1,479 @@
+                        ->exec();
+
+    if (response == QMessageBox::Yes) {
+        if (!m_list.deleteSkin(m_selected_skin, true)) {
+            m_list.deleteSkin(m_selected_skin, false);
+        }
+    }
+}
+
+void SkinManageDialog::on_urlBtn_clicked()
+{
+    auto url = QUrl(ui->urlLine->text());
+    if (!url.isValid()) {
+        CustomMessageBox::selectable(this, tr("Invalid url"), tr("Invalid url"), QMessageBox::Critical)->show();
+        return;
+    }
+
+    NetJob::Ptr job{ new NetJob(tr("Download skin"), APPLICATION->network()) };
+    job->setAskRetry(false);
+
+    auto path = FS::PathCombine(m_list.getDir(), url.fileName());
+    job->addNetAction(Net::Download::makeFile(url, path));
+    ProgressDialog dlg(this);
+    dlg.execWithTask(job.get());
+    SkinModel s(path);
+    if (!s.isValid()) {
+        CustomMessageBox::selectable(this, tr("URL is not a valid skin"),
+                                     QFileInfo::exists(path) ? tr("Skin images must be 64x64 or 64x32 pixel PNG files.")
+                                                             : tr("Unable to download the skin: '%1'.").arg(ui->urlLine->text()),
+                                     QMessageBox::Critical)
+            ->show();
+        QFile::remove(path);
+        return;
+    }
+    ui->urlLine->setText("");
+    if (QFileInfo(path).suffix().isEmpty()) {
+        QFile::rename(path, path + ".png");
+    }
+}
+
+class WaitTask : public Task {
+   public:
+    WaitTask() : m_loop(), m_done(false){};
+    virtual ~WaitTask() = default;
+
+   public slots:
+    void quit()
+    {
+        m_done = true;
+        m_loop.quit();
+    }
+
+   protected:
+    virtual void executeTask()
+    {
+        if (!m_done)
+            m_loop.exec();
+        emitSucceeded();
+    };
+
+   private:
+    QEventLoop m_loop;
+    bool m_done;
+};
+
+void SkinManageDialog::on_userBtn_clicked()
+{
+    auto user = ui->urlLine->text();
+    if (user.isEmpty()) {
+        return;
+    }
+    MinecraftProfile mcProfile;
+    auto path = FS::PathCombine(m_list.getDir(), user + ".png");
+
+    NetJob::Ptr job{ new NetJob(tr("Download user skin"), APPLICATION->network(), 1) };
+    job->setAskRetry(false);
+
+    auto uuidOut = std::make_shared<QByteArray>();
+    auto profileOut = std::make_shared<QByteArray>();
+
+    auto uuidLoop = makeShared<WaitTask>();
+    auto profileLoop = makeShared<WaitTask>();
+
+    auto getUUID = Net::Download::makeByteArray("https://api.mojang.com/users/profiles/minecraft/" + user, uuidOut);
+    auto getProfile = Net::Download::makeByteArray(QUrl(), profileOut);
+    auto downloadSkin = Net::Download::makeFile(QUrl(), path);
+
+    connect(getUUID.get(), &Task::aborted, uuidLoop.get(), &WaitTask::quit);
+    connect(getUUID.get(), &Task::failed, uuidLoop.get(), &WaitTask::quit);
+    connect(getProfile.get(), &Task::aborted, profileLoop.get(), &WaitTask::quit);
+    connect(getProfile.get(), &Task::failed, profileLoop.get(), &WaitTask::quit);
```

Comment:
I think the `failed` events should use some logging and or inform the user of the failed task